libraries/adaptive-expressions/src/triggerTrees/clause.ts

```diff
@@ -0,0 +1,285 @@
+                        if (shorterRel === RelationshipType.generalizes
+                            || (shorterRel === RelationshipType.specializes && shorterCount === longerCount)
+                            || shorterRel === RelationshipType.equal) {
+                            soFar = shorterRel;
+                        } else {
+                            break;
+                        }
+                    } else if (soFar != shorterRel) {
+                        // Not continued with sub/super so incomparable
+                        break;
+                    }
+                }
+            }
+
+            if (shorterCount !== longerCount) {
+                switch (soFar) {
+                    case RelationshipType.equal:
+                    case RelationshipType.generalizes:
+                        soFar = RelationshipType.generalizes;
+                        break;
+                    default:
+                        soFar = RelationshipType.incomparable;
+                        break;
+                }
+            }
+
+            soFar = this.bindingRelationship(soFar, shorter, longer);
+        }
+
+        return this.swap(soFar, swapped);
+    }
+
+    /**
+     * Determines whether the current `Clause` matches with another `Clause`.
+     * @param clause The other `Clause` instance to compare with.
+     * @param memory The scope for looking up variables.
+     * @returns A boolean value indicating whether the two clauses are matches.
+     */
+    public matches(clause: Clause, memory: any): boolean {
+        let matched = false;
+        if (clause.deepEquals(this)) {
+            matched = true;
+            if (this._ignored) {
+                const { value: match, error } = this._ignored.tryEvaluate(memory);
+                matched = !error && match;
+            }
+        }
+
+        return matched;
+    }
+
+    /**
+     * Splits ignored child expressions.
+     */
+    public splitIgnores(): void {
+        const children: Expression[] = [];
+        const ignores: Expression[] = [];
+        for (let i = 0; i < this.children.length; i++) {
+            const child = this.children[i];
+            if (child.type === ExpressionType.Ignore) {
+                ignores.push(child);
+            } else {
+                children.push(child);
+            }
+        }
+
+        this.children = children;
+
+        if (ignores.length > 0) {
+            this._ignored = Expression.andExpression(...ignores);
+        }
+    }
+
+    private bindingRelationship(soFar: RelationshipType, shorterClause: Clause, longerClause: Clause): RelationshipType {
+        if (soFar === RelationshipType.equal) {
+            let swapped = false;
+            let shorter = shorterClause.anyBindings;
+            let longer = longerClause.anyBindings;
+            if (Object.entries(shorterClause.anyBindings).length > Object.entries(longerClause.anyBindings).length) {
```

Comment:
If `anyBindings` was a `Map` type the `.size` can be read in constant time rather than linear. It also might make iteration cleaner.

Reply:
Changed `anyBindings` to `Map`.